### PR TITLE
114 get users location lobby

### LINKED ISSUE #114: get users location before show games in lobby

### PR DESCRIPTION
**Pull request by tamari-gray**
fixes #114 

- getting users location to use for distance calculations in #60

### CODE REVIEW

File: lib/screens/lobby/lobby.dart [hunk 1]
```diff
@@ -18,9 +22,9 @@ class LobbyScreen extends StatelessWidget {
           FlatButton(
             key: Key('signOutBtn'),
             onPressed: () {
-              // create a function to call on confirmation
+              // create a function to call and sign out on confirmation
               final signOut = () async {
-                context.read<Navigation>().pop();
+                Navigator.of(context).pop();
```
Comment (tamari-gray) [inline lib/screens/lobby/lobby.dart]: oops, gotta change that back

File: lib/screens/lobby/lobby.dart [hunk 2]
```diff
@@ -34,27 +38,22 @@ class LobbyScreen extends StatelessWidget {
...
-              }
-            }),
-      ),
+      body: FutureBuilder<Position>(
+          future: context.watch<LocationService>().getUsersCurrentLocation,
+          builder: (context, snapshot) {
+            if (snapshot.hasData == false) {
+              return Loading();
+            } else {
+              final userLocation = Location(
```
Comment (tamari-gray) [inline lib/screens/lobby/lobby.dart]: location will be passed to list of created games to do distance calculations in #60

File: lib/services/location_service.dart [hunk 3]
```diff
@@ -11,7 +11,7 @@ class LocationService {
       _geolocator.checkPermission();
 
   // get users currentLocation
-  Future<Position> getUsersCurrentLocation() =>
+  Future<Position> get getUsersCurrentLocation =>
```
Comment (tamari-gray) [inline lib/services/location_service.dart]: just a wee update

File: test/create_game_section/Lobby/lobby_widget_test.dart [hunk 4]
```diff
@@ -7,16 +7,64 @@ import 'package:flutter_test/flutter_test.dart';
...
+
+    // get users location
+    await tester.pumpAndSettle();
+
+    // show list of created games
+    expect(find.byType(ListOfCreatedGames), findsOneWidget);
+  });
   testWidgets(
       'shows loading icon + redirects to welcome screen on successfull logout',
       (WidgetTester tester) async {
```
Comment (tamari-gray) [inline test/create_game_section/Lobby/lobby_widget_test.dart:70]: all added code under this comment was to fix existing test.
(sorry if im just pointing out the obvious)

File: lib/screens/lobby/lobby.dart [hunk 5]
```diff
@@ -34,27 +38,22 @@ class LobbyScreen extends StatelessWidget {
           )
         ],
       ),
-      body: Container(
```
Comment (tamari-gray) [inline lib/screens/lobby/lobby.dart:37]: extracted red block below into its own file

**Review by tamari-gray — COMMENTED** [review]
woohoo shes ready for a review

File: lib/screens/lobby/list_of_created_games.dart [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  @override
+  Widget build(BuildContext context) {
+    // build list of tiles that show a game's info
+    return StreamBuilder<List<Game>>(
+        stream: context.watch<DatabaseService>().streamOfCreatedGames,
+        builder: (context, snapshot) {
+          // show empty screen if stream hasnt come through yet
+          if (snapshot.data == null) {
+            return Container(
+              key: Key('list_of_created_games_empty_screen'),
```
Comment (nickmeinhold) [inline lib/screens/lobby/list_of_created_games.dart:17]: Nice!  I never thought to give the empty state a key so you can check for it... love it!

File: lib/screens/lobby/list_of_created_games.dart [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+              child: ListView.builder(
+                itemCount: snapshot.data.length,
+                itemBuilder: (context, index) =>
+                    GameTile(snapshot.data[index], index),
+              ),
+            );
+          }
+        });
+  }
+}
```
Comment (nickmeinhold) [inline lib/screens/lobby/list_of_created_games.dart:32]: I liked this whole file - nice clean code I thought 👍

File: lib/screens/lobby/lobby.dart [hunk 8]
```diff
@@ -1,12 +1,16 @@
 import 'package:flutter/material.dart';
-import 'package:niira/models/game.dart';
-import 'package:niira/navigation/navigation.dart';
-import 'package:niira/screens/lobby/created_game_tile.dart';
+import 'package:geolocator/geolocator.dart';
```
Comment (nickmeinhold) [inline lib/screens/lobby/lobby.dart]: I reckon you're be better off removing the geolocator dependency - it's a clean architecture thing, happy to have a crack at explaining the benefits if you want, not sure if I'll do a good job or not. I can point you to a good reference if that's more helpful. Uncle Bob's books would be the best source I'd say, it's mainly the 'D' in the SOLID principles that's relevant here I'd say, off the top of my head... "Depend on abstractions, not on concretions."  [SOLID Principles made easy | Hacker Noon](https://hackernoon.com/solid-principles-made-easy-67b1246bcdf)

Comment (nickmeinhold) [reply lib/screens/lobby/lobby.dart]: So you'd have to make the LocationService convert (the geolocator) Position to (your homemade) Location objects so your return types can use Location (or something else that is your own PODO)

Comment (nickmeinhold) [reply lib/screens/lobby/lobby.dart]: But it's your call of course!

File: lib/screens/lobby/lobby.dart (hunk 5 above)
Comment (nickmeinhold) [reply lib/screens/lobby/lobby.dart:37]: nice!

File: lib/screens/lobby/lobby.dart (hunk 2 above)
Comment (nickmeinhold) [reply lib/screens/lobby/lobby.dart]: sweet!

File: lib/services/location_service.dart (hunk 3 above)
Comment (nickmeinhold) [reply lib/services/location_service.dart]: I think standard practice for a getter is to not have 'get' in the name.  I believe getters are conceptually the same as accessing a member and so should just be the member name.

File: test/mocks/services/mock_location_service.dart [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import 'dart:async';
+
+import 'package:geolocator/geolocator.dart';
+import 'package:mockito/mockito.dart';
+import 'package:niira/models/game.dart';
+import 'package:niira/models/location.dart';
+import 'package:niira/services/location_service.dart';
+
+class MockLocationService extends Mock implements LocationService {
```
Comment (nickmeinhold) [inline test/mocks/services/mock_location_service.dart]: If you are overriding the methods of the class that suggests to me that the mock might be better as a fake or a real object. What do you think?

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Hey mate!  Awesome work!  Had a couple of suggestions and questions but let me know if you'd prefer to keep moving and I'll hit the Approve button 👍

File: lib/screens/lobby/list_of_created_games.dart (hunk 7 above)
Comment (tamari-gray) [reply lib/screens/lobby/list_of_created_games.dart:32]: thank you :pray: 🙏

File: lib/screens/lobby/lobby.dart (hunk 8 above)
Comment (tamari-gray) [reply lib/screens/lobby/lobby.dart]: thanks for pointing it out! Super happy to implement that 😄

File: test/mocks/services/mock_location_service.dart (hunk 9 above)
Comment (tamari-gray) [reply test/mocks/services/mock_location_service.dart]: i like it, just refreshed my brain with mockito docs and im super keen to change it to a fake :)

File: lib/services/location_service.dart (hunk 3 above)
Comment (tamari-gray) [reply lib/services/location_service.dart]: sweet thank you!

**Comment by tamari-gray** [comment]
thanks for the SOLID review @nickmeinhold!
i added some changes and cleaned up her architecture 👷 🏗️ (think uncle bob will be proud of us cutting out that Position object)
shes all ready for another look (minor changes), when you are. thank you!

**Review by nickmeinhold — APPROVED** [review]
Bloody good work mate... and so fast 💨

**Comment by nickmeinhold** [comment]
oh and just a quick appreciation 👏 for the comedy... absolutely lovin the wordplay!

**Pull request merged**